pilot/pkg/bootstrap/options.go

```diff
@@ -136,8 +136,13 @@ var serviceAccountVar = env.RegisterStringVar("SERVICE_ACCOUNT", "", "")
 
 var revisionVar = env.RegisterStringVar("REVISION", "", "")
 
-// Apply default value to PilotArgs
-func (p *PilotArgs) Default() {
+// NewPilotArgs constructs pilotArgs with default values.
+func NewPilotArgs(initFunc func(*PilotArgs)) *PilotArgs {
+	p := &PilotArgs{}
+
+	// Apply custom initialization function first and apply defaults later.i
+	initFunc(p)
+
```

Comment:
Generally speaking customization should be applied on top of defaults. Otherwise we risk setting a field that the caller had intentionally cleared.

Also, you could generalize this function to take an arbitrary number of functions and remove the for loop on line 108.

```golang
bootstrap.NewPilotArgs(fns ..func(*bootstrap.PilotArgs) (*bootstrap.PilotArgs) {
    p := &PilotArgs{}    
    apply_defaults(p)    
    // apply any overrides
    for fn := range []fns {
        fn(p)    
    }
    return p
}
```

Reply:
Sorry, which for loop are you referring to in line 108?

Reply:
https://github.com/istio/istio/pull/20988/files#diff-5d62e3466efb64cae30503e344fe9a35R109-R111